server/push_notifications.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"fmt"
+	"github.com/mattermost/mattermost/server/public/model"
+)
+
+func (p *Plugin) NotificationWillBePushed(notification *model.PushNotification, userID string) (*model.PushNotification, string) {
+	// We will use our own notifications if:
+	// 1. This is a call start post
+	// 2. We have enabled ringing
+	// 3. The channel is a DM or GM
+	if notification.PostType != callStartPostType || !*p.getConfiguration().EnableRinging {
+		return nil, ""
+	}
+
+	if notification.ChannelType == model.ChannelTypeDirect || notification.ChannelType == model.ChannelTypeGroup {
+		return nil, "calls plugin will handle this notification"
+	}
+
+	// If it's a regular channel, then the user must have notifications set to all.
+	// In that case, make the notification nicer.
+	if notification.IsIdLoaded {
+		notification.Message = buildGenericPushNotificationMessage()
+		return notification, ""
+	}
+
+	nameFormat := p.getNotificationNameFormat(userID)
+	sender, appErr := p.API.GetUser(notification.SenderId)
+	if appErr != nil {
+		return nil, ""
```

Comment:
Can we log something? Unlikely but you never know :)

Reply:
Sure, of course. :)